src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs

```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
 
-            BoundExpression result = (temps.Count == 0 && stores.Count == 0) ?
+                return (temps.Count == 0 && stores.Count == 0) ?
                     conditionalExpression :
                     new BoundSequence(
                         syntax,
-                        temps.ToImmutable(),
-                        stores.ToImmutable(),
+                        temps.ToImmutableAndFree(),
+                        stores.ToImmutableAndFree(),
                         conditionalExpression,
                         conditionalExpression.Type);
+            }
 
-            temps.Free();
-            stores.Free();
+            // Rewrites the null coalescing operator in the case where the result type is the underlying
+            // non-nullable value type of the left side
+            BoundExpression rewriteNullCoalescingAssignmentForValueType()
+            {
+                Debug.Assert(node.LeftOperand.Type.IsNullableType());
+                Debug.Assert(node.Type.Equals(node.RightOperand.Type));
 
-            return result;
+                // We lower the expression to this form:
+                //
+                // var tmp = lhsRead.GetValueOrDefault();
+                // if (!lhsRead.HasValue) { tmp = loweredRight; transformedAssignment = tmp; }
+                // tmp
+                //
+                // Except that a is only evaluated once.
+
+                var leftOperand = node.LeftOperand;
+                if (!TryGetNullableMethod(leftOperand.Syntax,
+                                          leftOperand.Type,
+                                          SpecialMember.System_Nullable_T_GetValueOrDefault,
+                                          out var getValueOrDefault))
+                {
+                    return BadExpression(node);
+                }
+
+                if (!TryGetNullableMethod(leftOperand.Syntax,
+                                          leftOperand.Type,
+                                          SpecialMember.System_Nullable_T_get_HasValue,
+                                          out var hasValue))
+                {
+                    return BadExpression(node);
+                }
+
+                // If null coalescing assignment is supported in expression trees, the below code
+                // will need to be updated to support property accesses as well as calls. Currently,
+                // MakeRValue will never return a BoundPropertyAccess except in expression trees.
+                Debug.Assert(!_inExpressionLambda && lhsRead.Kind != BoundKind.PropertyAccess);
+
+                // If lhsRead is a call, such as to a property accessor, save the result off to a temp. This doesn't affect
+                // the standard ??= case because it only uses lhsRead once.
+                if (lhsRead.Kind == BoundKind.Call)
```

Comment:
>if (lhsRead.Kind == BoundKind.Call) [](start = 16, length = 35)

I didn't understand this. I would have expected `TransformCompoundAssignmentLHS` to  already have done what's needed, and we can just use `transformedLHS` that it returned.
FWIW, other scenarios that use `TransformCompoundAssignmentLHS` do not special-case calls.

Reply:
Correct, because other compound assignments do not have to read the value twice like we have to. It leaves calls to property accessors as is, so if we don't store the value to a temp we'll end up calling the property accessor twice.

---
In reply to: [293093678](https://github.com/dotnet/roslyn/pull/36329#discussion_r293093678) [](ancestors = 293093678)